docs/extras/simp.md

```diff
@@ -0,0 +1,60 @@
+# Simp #
+
+The `simp` tactic works by applying "a conditional term rewriting system" (I don't really know what this means and would be interested to know more), to try and simplify the goal. You can actually watch what it's doing (which to the untrained eye looks like trying to apply lots of lemmas tagged with simp until it either makes progress or stops) by writing `set_option trace.simplify true` in your code. If you do this in the below
+
+```lean
+namespace xena
+
+definition cong (a b m : ℤ) : Prop := ∃ n : ℤ, b - a = m * n
+
+notation a ` ≡ ` b ` mod ` m  := cong a b m 
+set_option trace.simplify true
+theorem cong_refl (m : ℤ) : ∀ a : ℤ, a ≡ a mod m :=
+begin
+intro a,
+unfold cong,
+existsi (0:ℤ),
+simp
+end
+
+end xena
+```
+
+you will discover that Lean seems to apply simp when you do an unfold as well (at least I am assuming that this is what's going on -- unfold is also underlined in green).
+
+If you only want to see what worked rather than all the things that didn't, you could try `set_option trace.simplify.rewrite true`.
+
+### Simp lemmas
+
+In case you want to train simp to use certain extra lemmas (for example because they're coming up again and again in your work) you can add new lemmas for yourself. For example in mathlib in `algebra/field.lean` we find the line
+
+```lean
+@[simp] theorem ne_zero (u : units α) : (u : α) ≠ 0
+```
+
+This lemma is then added to `simp`'s armoury. Note several things however.
+
+1) You can't just make a random theorem into a simp lemma. It has to be of a certain kind, the most important kinds being those of the form `A=B` and `A↔B`. Note however that if you want to add `fact` to `simp`'s weaponry, you can prove
+
+```lean
+@[simp] lemma my_lemma : fact <-> true
+```
+
+2) If you are not careful you can add a bad simp lemma of the form `foo x y = [something mentioning foo]` and then `simp` will attempt to rewrite `foo` and then end up with another one, and attempt to rewrite that, and so on. This can be fixed by using `rw` instead of `simp`, or using the config option `{single_pass := tt}`.
+
+
+### When not to use simp
+
+Don't use simp in the middle of proofs. Use it to finish proofs. If you really need to simplify a goal, use simp, and then cut and paste the goal into your code and write `suffices : (simplified thing), by simpa [this]` or some such thing. This is really important because the behaviour of simp changes sometimes, and if you put simp in the middle of proofs then your code might randomly stop compiling and it will be hard to figure out why if you didn't write down the exact thing which simp used to be doing.
```

Comment:
This should make it clear that the reason for the proscription is not that `simp` will not work, but rather that it will be brittle. "When not to use simp" sounds like it will describe places where `simp` is ineffective, not places where it will work but leads to poor quality code. Perhaps this should be called a "simp anti-pattern" instead, for those who know what that means.